### PR TITLE
Replace imp with importlib for Python 3.12

### PR DESCRIPTION
**Pull request by dviererbe**
Fix: #130

### CODE REVIEW

**Comment by dviererbe** [comment]
Changed base to reflect the changes made so far. We can change this back once this PR is ready for merging

File: lib/python/mod_python/cgihandler.py [hunk 1]
```diff
@@ -88,22 +88,22 @@ def handler(req):
...
                 raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
 
-            # this executes the module
-            imp.load_module(module_name, fd, path, desc)
+            if spec is None:
+                raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
+
+            module = importlib.util.module_from_spec(spec)
+            sys.modules[module_name] = module
+            spec.loader.exec_module(module)
```
Comment (panlinux) [inline lib/python/mod_python/cgihandler.py:100]: This logic was based on https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly, right?

Comment (dviererbe) [reply lib/python/mod_python/cgihandler.py:100]: Yes

**Comment by panlinux** [comment]
I tried http://10.0.102.65/cgi-python/yaml.py, and I don't have a `yaml.py` file in that directory. The `yaml` module of course exists. What happened is that I got a 503 instead of a 404. I added this debugging:

```diff
--- /usr/lib/python3/dist-packages/mod_python/cgihandler.py.orig        2024-02-21 19:00:03.169659018 +0000
+++ /usr/lib/python3/dist-packages/mod_python/cgihandler.py     2024-02-21 19:00:06.205700421 +0000
@@ -95,6 +95,7 @@
             if spec is None:
                 raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
 
+            sys.stderr.write(f"spec should be good and is {spec}\n")
             module = importlib.util.module_from_spec(spec)
             sys.modules[module_name] = module
             spec.loader.exec_module(module)
```

And got this in the log:
```
spec should be good and is ModuleSpec(name='yaml', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f3b2c15bb90>, origin='/var/www/html/cgi-python/yaml.py')
[Wed Feb 21 18:58:50.387630 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler: Traceback (most recent call last):
[Wed Feb 21 18:58:50.387647 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler:   File "/usr/lib/python3/dist-packages/mod_python/apache.py", line 402, in HandlerDispatch\n    result = obj(req)\n             ^^^^^^^^
[Wed Feb 21 18:58:50.387650 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler:   File "/usr/lib/python3/dist-packages/mod_python/cgihandler.py", line 101, in handler\n    spec.loader.exec_module(module)
[Wed Feb 21 18:58:50.387653 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler:   File "<frozen importlib._bootstrap_external>", line 991, in exec_module
[Wed Feb 21 18:58:50.387655 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler:   File "<frozen importlib._bootstrap_external>", line 1128, in get_code
[Wed Feb 21 18:58:50.387657 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler:   File "<frozen importlib._bootstrap_external>", line 1186, in get_data
[Wed Feb 21 18:58:50.387660 2024] [:error] [pid 59134:tid 139892149409472] [client 10.0.102.1:43542] PythonHandler mod_python.cgihandler: FileNotFoundError: [Errno 2] No such file or directory: '/var/www/html/cgi-python/yaml.py'
```

**Comment by panlinux** [comment]
This is unfortunate :/
```python3
>>> spec = importlib.util.spec_from_file_location("yaml", "/does/not/exist.py")
>>> spec
ModuleSpec(name='yaml', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7fbe4d5fbaa0>, origin='/does/not/exist.py')
```
Looks like I don't understand what `spec_from_file_location()` should do.

**Comment by panlinux** [comment]
Maybe we should just catch that in `exec_module()`

**Comment by grisha** [comment]
So do I understand this correctly - the `if spec is None:` part doesn't work because spec exists even for non existing files. 

It looks like it will blow up later in `spec.loader.exec_module(module)`, but that's not an `HTTP_NOT_FOUND`, which would be nicer.

Of course we could just do this (unless I'm missing something):

```python
if not os.path.exists(os.path.join(dir, file)):
     raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
```

P.S. I don't think whether `yaml` exists matters, for example I don't have `yaml` on this machine and it looks like this:
```python
>>> import yaml
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'yaml'
>>> spec = importlib.util.spec_from_file_location("yaml", "/does/not/exist.py")
>>> spec
ModuleSpec(name='yaml', loader=<_frozen_importlib_external.SourceFileLoader object at 0x82bf518b0>, origin='/does/not/exist.py')
>>> module = importlib.util.module_from_spec(spec)
>>> spec.loader.exec_module(module)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<frozen importlib._bootstrap_external>", line 846, in exec_module
  File "<frozen importlib._bootstrap_external>", line 982, in get_code
  File "<frozen importlib._bootstrap_external>", line 1039, in get_data
FileNotFoundError: [Errno 2] No such file or directory: '/does/not/exist.py'
>>> 
```

**Comment by panlinux** [comment]
> Of course we could just do this (unless I'm missing something):
> 
> ```python
> if not os.path.exists(os.path.join(dir, file)):
>      raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
> ```

Yeah, that looks better. I worry about "dir" and "file" with shenaningans like `../../../../etc/passwd`, I hope something else sanitized things before this part of the code is reached...

File: lib/python/mod_python/cgihandler.py (hunk 1 above)
Comment (panlinux) [inline lib/python/mod_python/cgihandler.py:100]: Should we check if the module is imported already, before doing any of this? 

https://docs.python.org/3/library/importlib.html#checking-if-a-module-can-be-imported deals with such a case, but it's not in the example at https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly

Comment (dviererbe) [reply lib/python/mod_python/cgihandler.py:100]: > Should we check if the module is imported already, before doing any of this?

I think we do not want this as we want to execute the module on every call, right?

**Comment by dviererbe** [comment]
What do you think about this?

1. checking that the script exists
   ```python
   scriptPath = os.path.join(dir, file)
   
   if not os.path.exists(scriptPath):
       raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
   ```
2. checking that the script is contained in `dir` to avoid shenaningans like `../../../../etc/passwd`
   ```python
   scriptPath = os.path.abspath(scriptPath)
   if not scriptPath.startswith(dir):
       raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)
   ```
3. load module
   ```python
   try:
       # we do not search the pythonpath (security reasons)
       spec = importlib.util.spec_from_file_location(module_name, scriptPath)                                                   
   except (ModuleNotFoundError, ValueError):
       raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)

   if spec is None:
       raise apache.SERVER_RETURN(apache.HTTP_NOT_FOUND)

   module = importlib.util.module_from_spec(spec)
   sys.modules[module_name] = module
   spec.loader.exec_module(module)
 
   return apache.OK
   ```

**Comment by grisha** [comment]
Looks good to me!

**Comment by dviererbe** [comment]
The conflict arises because I squashed all changes.

**Comment by grisha** [comment]
I'll let it be for a few days to make sure no other problems are discovered, then merge it into master. Thank you @dviererbe and @panlinux !

**Pull request merged**